pkg/injector/bootstrap.go

```diff
@@ -100,3 +105,103 @@ func (wh *mutatingWebhook) marshalAndSaveBootstrap(name, namespace string, confi
 	log.Debug().Msgf("Creating bootstrap config for Envoy: name=%s, namespace=%s", name, namespace)
 	return wh.kubeClient.CoreV1().Secrets(namespace).Create(context.Background(), secret, metav1.CreateOptions{})
 }
+
+// NewBootstrapSecretRotator returns a new bootstrap secret rotator.
+func NewBootstrapSecretRotator(ctx context.Context, kubeClient kubernetes.Interface, informerCollection *informers.InformerCollection, certManager *certificate.Manager, checkInterval time.Duration) *BootstrapSecretRotator {
+	return &BootstrapSecretRotator{
+		context:            ctx,
+		kubeClient:         kubeClient,
+		informerCollection: informerCollection,
+		certManager:        certManager,
+		checkInterval:      checkInterval,
+	}
+}
+
+// listBootstrapSecrets returns the bootstrap secrets stored in the informerCollection's store.
+// this function is used to get the namespace of the secrets.
+func (b *BootstrapSecretRotator) listBootstrapSecrets() []*corev1.Secret {
+	// informers return slice of pointers so we'll convert them to value types before returning
+	secretPtrs := b.informerCollection.List(informers.InformerKeySecret)
+	var secrets []*corev1.Secret
+
+	for _, secretPtr := range secretPtrs {
+		if secretPtr == nil {
+			continue
+		}
+		secret, ok := secretPtr.(*corev1.Secret)
+		if !ok {
+			continue
+		}
+		// finds bootstrap secrets
+		if strings.Contains(secret.Name, bootstrapSecretPrefix) {
+			secrets = append(secrets, secret)
+		}
+	}
+
+	return secrets
+}
+
+// rotateBootstrapSecrets finds the bootstrap secret of the given certificate
+// from the list of secrets stored in the informerCollection's store and updates the secret.
+func (b *BootstrapSecretRotator) rotateBootstrapSecrets() {
+	secrets := b.listBootstrapSecrets()
+	for _, secret := range secrets {
+		secretProxyUUID := strings.ReplaceAll(secret.Name, bootstrapSecretPrefix, "")
+		certs := b.certManager.ListIssuedCertificates()
+		for _, cert := range certs {
```

Comment:
alternatively, we can query the proxyRegistry, by UUID, and reissue the certificate. Remember that the ListIssuedCertificates only has a cache, and may not actually have the certificate here

Reply:
this is also currently an O(MXN) operation (roughly N^2 if everything is held in memory), which we can reduce to O(N) (assuming constant lookup on proxy registry)

Reply:
If we use `ListIssuedCertificate()` and the cert cache is empty or doesn't contain the bootstrap cert then it will never be rotated even though it is expiring. 

I think this is one potential solution:
1. Iterate over the bootstrap secrets
2. Check if the bootstrap cert is going to expire. We could call `IssueCertificate` here first but there is a high probability that the cert is not in the local cache and therefore a new cert would be returned. In this case, we'd be updating the secret a lot more frequently. Perhaps there is some other check that we could perform to prevent this and simply call `IssueCertificate` first.
3. If the cert is going to expire, call `IssueCertificate()`
4. Update the secret with the new cert
5. If the update fails, log the error and continue

Reply:
> alternatively, we can query the proxyRegistry, by UUID, and reissue the certificate. Remember that the ListIssuedCertificates only has a cache, and may not actually have the certificate here

This check could then be moved to the osm-controller from the osm-injector

Reply:
>We could call IssueCertificate here first but there is a high probability that the cert is not in the local cache

Why is it a high probability?

Reply:
> Check if the bootstrap cert is going to expire. 

This makes sense to me.  If I am following correctly, this currently is only rotating when a cert is in the local cache. If its not in the local cache we would potentially miss it.

How are we responding to Root Certificate rotation?  Right now, If the root rotates in but the cert isn't expiring then we wouldn't update it.

Seems like there are two times we want to rotate the bootstrap:  

- if we are close to expiration
- if new root rotates in

Reply:
> > We could call IssueCertificate here first but there is a high probability that the cert is not in the local cache
> 
> Why is it a high probability?

Good point, we can't assume this is a high probability. I guess what we can say is it is a possibility so we can't rely on the cache alone.

Reply:
> > Check if the bootstrap cert is going to expire.
> 
> This makes sense to me. If I am following correctly, this currently is only rotating when a cert is in the local cache. If its not in the local cache we would potentially miss it.
> 
> How are we responding to Root Certificate rotation? Right now, If the root rotates in but the cert isn't expiring then we wouldn't update it.
> 
> Seems like there are two times we want to rotate the bootstrap:
> 
> * if we are close to expiration
> * if new root rotates in

Calling `IssueCertificate()` will check if the cert (if present in the cache) needs to be rotated due to root certificate rotation by comparing the issuing and validating IDs to the issuers in the `certificate.Manager`. You're correct that if we pre-check for expiration we also need to pre-check for mismatched IDs.

Reply:
If the likelihood of injector churn is low enough, then I think calling `IssueCertificate()` without a pre check would be acceptable.

Reply:
I think you are right that there needs to be some type of check. I think `listBootstrapSecrets()` iterates over every secret in the cluster? If there are multiple instances of the injector, we would be potential calling this for every secret, sometime multiple times across the instances.

Reply:
Yeah, that's correct. I am not sure how we could divide secret/certificate updates across the injectors so there is no overlap.

Reply:
Maybe this process would make more sense in the osm-bootstrap since there are likely to be less replicas. @steeling what are your thoughts on this?

Reply:
I guess this problem might be good justification for using the ProxyRegistry. We wouldn't see the same overlap using the proxyRegistry and moving this process to the osm-controller If we construct the certificate names from the proxies in the registry we would still need to know the namespace to look up the Kubernetes secret.

Additionally, if we use the proxyRegistry to lookup bootstrap secrets/certificates and a proxy is failing to connect to the control plane for whatever reason its bootstrap certificate may never be updated it isn't in the proxyRegistry and therefore may never be able to connect to the control plane.

Reply:
if we implement using the ProxyRegistry then I don't think the code should be in the pkg/injector folder, but not sure where would be a better place, any ideas? @jaellio @jsturtevant @steeling

Reply:
@jaellio it's quite possible we're going to remove the bootstrap pod with https://github.com/openservicemesh/osm/issues/4967, so let's hhold off on that.

Reply:
@steeling If we continue iterating over the secrets directly (without the proxyRegistry) how could we limit the high churn of multiple injectors trying to update all bootstrap secrets at the same time? Unless they all check the same certs in exactly the same order and at the same time over and over again I think they will eventually be updated. I just don't want to run into a similar issue that we are seeing with #4845

Reply:
an update wouldn't change the owner reference nor create a secret that was previously deleted, so won't result in https://github.com/openservicemesh/osm/issues/4845.

I'd say that secrets should be rotated infrequently enough that attempting to solve this issue now would be a premature optimization